frame/contracts/src/exec.rs

```diff
@@ -168,91 +171,110 @@ pub trait Ext {
 
 	/// Returns the price for the specified amount of weight.
 	fn get_weight_price(&self, weight: Weight) -> BalanceOf<Self::T>;
+
+	/// Get a reference to the schedule used by the current call.
+	fn schedule(&self) -> &Schedule<Self::T>;
 }
 
-/// Loader is a companion of the `Vm` trait. It loads an appropriate abstract
-/// executable to be executed by an accompanying `Vm` implementation.
-pub trait Loader<T: Config> {
-	type Executable;
-
-	/// Load the initializer portion of the code specified by the `code_hash`. This
-	/// executable is called upon instantiation.
-	fn load_init(&self, code_hash: &CodeHash<T>) -> Result<Self::Executable, &'static str>;
-	/// Load the main portion of the code specified by the `code_hash`. This executable
-	/// is called for each call to a contract.
-	fn load_main(&self, code_hash: &CodeHash<T>) -> Result<Self::Executable, &'static str>;
+/// Describes the different functions that can be exported by an [`Executable`].
+pub enum ExportedFunction {
+	/// The constructor function which is executed on deployment of a contract.
+	Constructor,
+	/// The function which is executed when a contract is called.
+	Call,
 }
 
-/// A trait that represent a virtual machine.
+/// A trait that represents something that can be executed.
 ///
-/// You can view a virtual machine as something that takes code, an input data buffer,
-/// queries it and/or performs actions on the given `Ext` and optionally
-/// returns an output data buffer. The type of code depends on the particular virtual machine.
-///
-/// Execution of code can end by either implicit termination (that is, reached the end of
-/// executable), explicit termination via returning a buffer or termination due to a trap.
-pub trait Vm<T: Config> {
-	type Executable;
+/// In the on-chain environment this would be represented by a wasm module. This trait exists in
+/// order to be able to mock the wasm logic for testing.
+pub trait Executable<T: Config>: Sized {
+	/// Load the executable from storage.
+	fn from_storage(code_hash: CodeHash<T>, schedule: &Schedule<T>) -> Result<Self, DispatchError>;
+
+	/// Load the module from storage without re-instrumenting it.
+	///
+	/// A code module is re-instrumented on-load when it was originally instrumented with
+	/// an older schedule. This skips this step for cases where the code storage is
+	/// queried for purposes other than execution.
+	fn from_storage_noinstr(code_hash: CodeHash<T>) -> Result<Self, DispatchError>;
+
+	/// Put the executable into storage. If it already exists there the refcount is incremented.
+	fn store(self);
+
+	/// Decrements the refcount by one and deletes the code if it drops to zero.
+	fn store_decremented(self);
```

Comment:
I kind of feel like `store_code` and `drop_code` would be more self-explanatory or maybe `store_executable` and `drop_executable`. But feel free to ignore this suggestion if it is not convincing

Reply:
Yeah the names are bad. You are the second person who stumbles over them. Let me think about that.